File: backend/app/schemas/trade.py
from pydantic import BaseModel
from typing import List, Optional
from datetime import date, datetime


class TradeBase(BaseModel):
    portfolio_code: str
    product_code: str
    market: Optional[str] = None
    platform_code: Optional[str] = None
    trade_type: str
    transfer_group: Optional[str] = None  # 平台间现金转移配对标识
    shares: Optional[float] = None
    amount: Optional[float] = None
    price: Optional[float] = None
    fee: float = 0
    actual_amount: Optional[float] = None
    trade_date: date
    confirm_date: Optional[date] = None
    status: str = "pending"
    notes: Optional[str] = None


class TradeCreate(TradeBase):
    # 「实际金额」= 一笔基金交易真正进出账户的现金（方向敏感）：
    #   买入 = 含费现金支出（资金流出，含手续费）；卖出 = 到手净额（资金流入，已扣手续费）。
    # 输入层 amount / actual_amount 两参同义、均指「实际金额」，actual_amount 优先；
    # 买入按 actual_amount − fee 联动落库；卖出为纯派生量（见下）。
    # 卖出金额为纯派生量（#190）：有价格时 amount = shares × price、
    #   actual_amount = amount − fee；显式传入的「实际金额」仅作一致性校验
    #   （差值超 0.01 报 AMOUNT_MISMATCH），落库恒用推导值。
    # 自然键防重（#82）：命中同参数 pending/confirmed 交易时需显式传 True 强制创建
    allow_duplicate: bool = False
    # 跨平台现金腿（#91）：买=扣款平台、卖=到账平台，缺省同基金腿平台
    cash_platform_code: Optional[str] = None
    # #93: CASH 腿独立确认日（卖出到账日）
    cash_confirm_date: Optional[date] = None


class TradeUpdate(BaseModel):
    # confirm_date 不开放直改：创建/unconfirm 按 confirm_days 自动维护，
    # 改 trade_date 时由 service 联动重算；补录覆盖走 confirm 端点传参
    # status 不开放直改：状态流转走 confirm/cancel/unconfirm 端点
    # amount 语义（#182 D1，与创建同口径）：「实际金额」= 真正进出账户的现金
    # （买入=含费现金支出、卖出=到手净额）；amount/actual_amount 两参同义、
    # actual_amount 优先，service 联动重算净额列并镜像 CASH 腿
    shares: Optional[float] = None
    amount: Optional[float] = None
    price: Optional[float] = None
    fee: Optional[float] = None
    actual_amount: Optional[float] = None
    trade_date: Optional[date] = None
    notes: Optional[str] = None
    # #493：已确认卖出的到账日修正入口（只同步配对 CASH 腿的 confirm_date，
    # 不重算基金金额）。仅 confirmed + sell 接受该字段；pending 传它报
    # INVALID_PARAM、显式 null 拒绝、混入其他财务字段整体拒绝——组合限制
    # 留在 service（schemas 只做透传）。
    cash_confirm_date: Optional[date] = None


class TradeResponse(TradeBase):
    id: int
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    # product_name 读侧派生（非 DB 列）：仅 list 端点批量 join 产品表填充，
    # create/get/update/preview 响应恒为 None；同 positions 模式（#175）
    product_name: Optional[str] = None
    # #493 只读派生（非 DB 列，不落库）：仅**基金腿**从实际配对 CASH 腿读取；
    # 无现金腿（如待确认卖出）时为 null，CASH 腿自身不回填。批量来源见
    # trade_service.build_paired_cash_leg_map
    cash_platform_code: Optional[str] = None
    cash_confirm_date: Optional[date] = None

    class Config:
        from_attributes = True


class PaginatedTradeResponse(BaseModel):
    """交易列表分页响应（issue #183）。items 元素复用 TradeResponse：
    product_name 仅 list 端点填充，单对象端点恒为 None（见 TradeResponse 注释）。"""
    items: List[TradeResponse]
    total: int
    page: int
    page_size: int


class TradePreviewResult(BaseModel):
    """确认前预览的计算结果（与真实确认共用同一计算实现）"""
    price: Optional[float] = None
    shares: Optional[float] = None
    amount: Optional[float] = None
    actual_amount: Optional[float] = None
    fee: float = 0
    confirm_date: Optional[date] = None
    nav_date: Optional[date] = None  # OTC 净值型时取净值的 T 日
    is_otc_nav_fund: bool = False
    # #493：本次确认将使用的**有效**现金平台/日期（买 = 扣款平台与 T；
    # 卖 = 到账平台与到账日 A，缺省 A=C）。与 `trade.cash_*`（实际已存在的
    # 配对 CASH 腿信息）刻意区分：确认前卖出还没有现金腿。
    cash_platform_code: Optional[str] = None
    cash_confirm_date: Optional[date] = None


class TradePreviewResponse(BaseModel):
    trade: TradeResponse
    preview: TradePreviewResult
    paired_cash_amount: Optional[float] = None  # 配对 CASH 腿将同步的金额

File: frontend/src/types/trade.ts
export interface Trade {
  id: number;
  portfolio_code: string;
  product_code: string;
  market?: string;
  platform_code?: string;
  trade_type: string;
  shares?: number;
  amount?: number;
  price?: number;
  fee: number;
  actual_amount?: number;
  trade_date: string;
  confirm_date?: string;
  status: string;
  /** 业务分组（#126 决策⑨）：rebal_*=调仓配对、sub_*=申赎现金腿、12位hex=现金转移；仅用于结对展示，页面不展示该编码 */
  transfer_group?: string;
  notes?: string;
  created_at?: string;
  updated_at?: string;
  /** 读侧派生（#175）：仅 list 响应有值；create/get/update 恒为 undefined */
  product_name?: string;
  /**
   * 读侧派生现金信息（#493）：**只读、不落库、仅基金腿填充**——后端按
   * `(portfolio_code, transfer_group)` 批量查反向 CASH 腿，因此不受列表的
   * 状态/平台筛选与分页截断影响，半确认组（买入扣款腿已 confirmed、基金腿
   * pending）同样可见；待确认卖出尚无 CASH 腿 → 两者为 null。
   * CASH 腿自身不回填（恒 null）。列表/详情一律读这两个字段，
   * **不要**再从当前页里寻找配对 CASH 行。
   * 语义：买入=扣款平台/扣款日（=下单日 T）；卖出=到账平台/到账日 A。
   */
  cash_platform_code?: string | null;
  cash_confirm_date?: string | null;
}

export interface TradeCreate {
  portfolio_code: string;
  product_code: string;
  market?: string;
  platform_code?: string;
  /**
   * 跨平台现金腿（#91/#493）：**仅买入可用**=扣款平台（缺省同基金腿平台），
   * 落库为配对 CASH sell 腿并创建即 confirmed。
   * 卖出传它 → 422 CASH_PLATFORM_NOT_ALLOWED（到账平台改在 confirm/preview 录入）。
   */
  cash_platform_code?: string;
  trade_type: string;
  shares?: number;
  amount?: number;
  price?: number;
  fee?: number;
  actual_amount?: number;
  trade_date: string;
  notes?: string;
  /** 命中 DUPLICATE_TRADE 时用户确认后重试传 true（后端默认 false） */
  allow_duplicate?: boolean;
}

// 字段与后端 schemas/trade.py::TradeUpdate 对齐：
// 不含 confirm_date/status（后端会静默丢弃）；改状态请走 confirm/unconfirm/cancel 端点
export interface TradeUpdate {
  shares?: number;
  // amount 语义（#182 D1，与创建同口径）：buy/sell 均为实际金额口径
  // （buy=含费现金支出、sell=到手净额），与 actual_amount 同义、后者优先
  amount?: number;
  price?: number;
  fee?: number;
  actual_amount?: number;
  trade_date?: string;
  notes?: string;
  /**
   * 卖出到账日修正（#493）：**已确认卖出的窄例外**——只允许
   * `{cash_confirm_date, notes?}` 子集，混入任何其他字段整体拒绝、显式 null 拒绝；
   * pending 交易传它 → 422 INVALID_PARAM（pending 走 trade_date 重算）。
   */
  cash_confirm_date?: string;
}

// 确认前预览（#248）：与真实确认共用后端计算实现
export interface TradePreviewResult {
  price?: number;
  shares?: number;
  amount?: number;
  actual_amount?: number;
  fee: number;
  confirm_date?: string;
  nav_date?: string;
  is_otc_nav_fund: boolean;
  /** 本次有效扣款/到账平台（#493）：买=扣款平台（创建期已定）、卖=传入值或基金腿平台 */
  cash_platform_code?: string | null;
  /** 本次有效现金日（#493）：买=下单日 T、卖=到账日 A（缺省 = 本次有效确认日 C） */
  cash_confirm_date?: string | null;
}

export interface TradePreviewResponse {
  trade: Trade;
  preview: TradePreviewResult;
  paired_cash_amount?: number;
}

/**
 * 确认/preview 的业务输入（#493）：两者同一组参数，**一律走 query 参数**
 * （后端 confirm/preview 端点均为 query 协议，JSON body 会被静默忽略）。
 * - `confirm_date`：有效基金确认日 C（缺省取创建时按 confirm_days 推导的值）
 * - `price`：场内成交价覆盖；场外仅作与 T 日净值的一致性校验
 * - `cash_confirm_date`：卖出到账日 A（缺省 = C；买入只接受等于 T）
 * - `cash_platform_code`：卖出到账平台（缺省 = 基金腿平台；买入不可改扣款平台）
 */
export interface TradeConfirmParams {
  confirm_date?: string;
  price?: number;
  cash_confirm_date?: string;
  cash_platform_code?: string;
  /** 仅 confirm 端点：命中 MISSING_NAV 时显式同步净值后重试（preview 不产生写入） */
  sync_nav?: boolean;
}

/**
 * 确认端点的外层响应结构（#493 §3.3：保持现状，新增信息进入其中的 `trade`）：
 * 顶层平铺字段是历史契约（CLI 也在读），**交易本体在 `trade` 字段里**。
 */
export interface TradeConfirmResponse {
  message: string;
  id: number;
  portfolio_code: string;
  trade_type: string;
  status: string;
  confirm_date?: string | null;
  trade: Trade;
}
